presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -219,47 +218,49 @@ protected RelationPlan visitJoin(Join node, Void context)
+                if (!isEqualComparisonExpression(conjunct) && node.getType() != INNER) {
                     throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins only supported for inner join: %s", conjunct);
                 }
-                Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(comparison.getLeft(), analysis.getColumnReferences());
-                Set<QualifiedName> secondDependencies = DependencyExtractor.extractNames(comparison.getRight(), analysis.getColumnReferences());
-
-                Expression leftExpression;
-                Expression rightExpression;
-                if (firstDependencies.stream().allMatch(left.canResolvePredicate()) && secondDependencies.stream().allMatch(right.canResolvePredicate())) {
-                    leftExpression = comparison.getLeft();
-                    rightExpression = comparison.getRight();
-                }
-                else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && secondDependencies.stream().allMatch(left.canResolvePredicate())) {
-                    leftExpression = comparison.getRight();
-                    rightExpression = comparison.getLeft();
-                    comparisonType = flipComparison(comparisonType);
+
+                if (conjunct instanceof ComparisonExpression) {
+                    Expression firstExpression = ((ComparisonExpression) conjunct).getLeft();
+                    Expression secondExpression = ((ComparisonExpression) conjunct).getRight();
+                    ComparisonExpression.Type comparisonType = ((ComparisonExpression) conjunct).getType();
+                    Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(firstExpression, analysis.getColumnReferences());
+                    Set<QualifiedName> secondDependencies = DependencyExtractor.extractNames(secondExpression, analysis.getColumnReferences());
+
+                    if (firstDependencies.stream().allMatch(left.canResolvePredicate()) && secondDependencies.stream().allMatch(right.canResolvePredicate())) {
+                        leftComparisonExpressions.add(firstExpression);
+                        rightComparisonExpressions.add(secondExpression);
+                        joinConditionComparisonTypes.add(comparisonType);
+                    }
+                    else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && secondDependencies.stream().allMatch(left.canResolvePredicate())) {
+                        leftComparisonExpressions.add(secondExpression);
+                        rightComparisonExpressions.add(firstExpression);
+                        joinConditionComparisonTypes.add(flipComparison(comparisonType));
+                    }
+                    else {
+                        // the case when we mix symbols from both left and right join side on either side of condition.
+                        nonComparisonExpressions.add(conjunct);
```

Comment:
This name doesn't seem right. The variable is named `nonComparisonExpressions` but the only expressions you add to it are ones that pass `conjunct instance of ComparisonExpression`. Maybe name it `complexComparisonExpressions`?